pkg/terraform/store.go

```diff
@@ -158,9 +158,22 @@ func (ws *WorkspaceStore) Workspace(ctx context.Context, c resource.SecretClient
 	if err := fp.EnsureTFState(ctx); err != nil {
 		return nil, errors.Wrap(err, "cannot ensure tfstate file")
 	}
+	isNeedProviderUpgrade, err := fp.isNeedProviderUpgrade()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot upgrade provider")
+	}
 	if err := fp.WriteMainTF(); err != nil {
 		return nil, errors.Wrap(err, "cannot write main tf file")
 	}
+	if isNeedProviderUpgrade {
+		cmd := w.executor.CommandContext(ctx, "terraform", "init", "-upgrade", "-input=false")
+		cmd.SetDir(w.dir)
+		out, err := cmd.CombinedOutput()
+		w.logger.Debug("init ended", "out", string(out))
```

Comment:
nit: We had better distinguish this debug log with the regular `terraform init`'s debug log to prevent ambiguity (if the Terraform CLI output already contains this information, then we can also rely on that):
```suggestion
		w.logger.Debug("init -upgrade ended", "out", string(out))
```